monero-oxide/ringct/bulletproofs/src/plus/weighted_inner_product.rs

```diff
@@ -341,29 +342,42 @@ impl WipStatement {
     };
 
     let mut e_is = Vec::with_capacity(proof.L.len());
-    for (L, R) in proof.L.iter_mut().zip(proof.R.iter_mut()) {
-      e_is.push(Self::transcript_L_R(&mut transcript, *L, *R));
-      *L = L.mul_by_cofactor();
-      *R = R.mul_by_cofactor();
+    let mut L = Vec::with_capacity(proof.L.len());
+    let mut R = Vec::with_capacity(proof.R.len());
+
+    let decomp_mul_cofactor = |p| decompress_point(p).map(|p| EdwardsPoint::mul_by_cofactor(&p));
+
+    for (L_i, R_i) in proof.L.into_iter().zip(proof.R.into_iter()) {
+      e_is.push(Self::transcript_L_R(&mut transcript, L_i, R_i));
+
+      let (Some(L_i), Some(R_i)) = (decomp_mul_cofactor(L_i), decomp_mul_cofactor(R_i)) else {
+        return false;
+      };
+
+      L.push(L_i);
+      R.push(R_i);
     }
 
     let e = Self::transcript_A_B(&mut transcript, proof.A, proof.B);
-    proof.A = proof.A.mul_by_cofactor();
-    proof.B = proof.B.mul_by_cofactor();
+
+    let (Some(A), Some(B)) = (decomp_mul_cofactor(proof.A), decomp_mul_cofactor(proof.B)) else {
```

Comment:
As with `Bulletproofs`, we can de-structure the point to ensure it's out-of-scope.